subsys/net/lib/lwm2m_client_utils/fota/settings.c

```diff
@@ -7,54 +7,76 @@
 
 #include <zephyr.h>
 #include <settings/settings.h>
+#include <stdlib.h>
+#include <stdio.h>
 
 #include <net/lwm2m_client_utils_fota.h>
 
 #include <logging/log.h>
 LOG_MODULE_REGISTER(lwm2m_fota_settings, CONFIG_LWM2M_CLIENT_UTILS_LOG_LEVEL);
 
-static struct update_counter uc;
+#define MAX_INSTANCE_COUNT CONFIG_LWM2M_FIRMWARE_INSTANCE_COUNT
 
-int fota_update_counter_read(struct update_counter *update_counter)
+static struct update_counter uc[MAX_INSTANCE_COUNT];
+
+static const char fctr_path[] = "fota/counter";
+static const char ctr_str[] = "counter";
+
+int fota_update_counter_read(uint16_t obj_inst_id, struct update_counter *update_counter)
 {
-	memcpy(update_counter, &uc, sizeof(uc));
+	memcpy(update_counter, &(uc[obj_inst_id]), sizeof(uc[obj_inst_id]));
 	return 0;
 }
 
-int fota_update_counter_update(enum counter_type type, uint32_t new_value)
+int fota_update_counter_update(uint16_t obj_inst_id, enum counter_type type, uint32_t new_value)
 {
+	char id[6];
+	char path[sizeof(fctr_path) + 1 + 5]; /* prefix + '/' + obj insta id digits(uint16_t)*/
```

Comment:
This is going to be an incompatibility issue with already released versions. The change will prevent updates from older releases and should be stated in the changelog. I assume this change could be merged only when there is a major version update. One workaround that comes to my mind is to use the prefix-only path for the first instance. I didn't try that with my version though.